### PR TITLE
availabilites test compeleted

### PR DESCRIPTION
**Pull request by Kwasi008**
Hi John I just finished the schedule window test, please let know what you think.

@micronix 
@jrmcgarvey 

Thank you.

### CODE REVIEW

File: app/controllers/api/v1/schedule_windows.rb [hunk 1]
```diff
@@ -44,6 +44,9 @@ class ScheduleWindows < Grape::API
         schedule = ScheduleWindow.where(id: permitted_params[:id])
         all_events = []
         create_all_events(schedule, Date.today, Date.today+1.year, all_events)
+        if params[:is_recurring] == false
+          return { seccuss:true}
```
Comment (jrmcgarvey) [inline app/controllers/api/v1/schedule_windows.rb]: Misspelling here.  I think you want success: true

It troubles me that no json is rendered for this case.  It doesn't look right.  This change actually changes the API.  I don't think it is correct.

File: spec/factories/window_schedule.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+    factory :schedule_window do
+   start_date {"02-01-2019"}
+    end_date {"05-03-2019"}
```
Comment (jrmcgarvey) [inline spec/factories/window_schedule.rb]: A typical schedule window would only be a couple of hours, not months.  Also, it is not clear how starttime and endtime are used as compared with startdate and enddate, so I am not sure these values are valid.

File: spec/factories/window_schedule.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+    start_time {"09-03-2019"}
+    end_time {"10-02-2019"} 
+    is_recurring {false}
```
Comment (jrmcgarvey) [inline spec/factories/window_schedule.rb]: We actually need two test cases here.  (They could be in separate PRs.) We need to test the case where is_recurring is false, but we also need to test the case where is_recurring is true.

File: spec/requests/schedule_windows_spec.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    let!(:organization) { FactoryBot.create(:organization) }
+    let!(:driver) { FactoryBot.create(:driver, organization_id: organization.id) }
+    let!(:location) { FactoryBot.create(:location) }
+    it "Sign up for availabilities" do 
+     post '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken}, params: { 
+         start_date: "02-01-2019",
+         end_date: "05-03-2019", 
+         start_time: "09-03-2019", 
+         end_time: "10-02-2019", 
+         is_recurring: true,
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: Here we have an is_recurring: true but we don't have a factory or anything for the recurring_pattern.  The problem here is not that the test case is written wrong.  It's that we don't understand how the API is supposed to work.  How can one specify is_recurring=true without specifying a recurring_pattern?

File: spec/requests/schedule_windows_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  
+    let!(:organization) { FactoryBot.create(:organization) }
+    let!(:driver) { FactoryBot.create(:driver, organization_id: organization.id) }
+    let!(:location) { FactoryBot.create(:location) }
+    let!(:schedule_window) { FactoryBot.create(:schedule_window, driver_id: driver.id, location_id: location.id) }
+    it "update availabilities" do 
+       put "/api/v1/availabilities/#{schedule_window.id}", headers: {"ACCEPT" => "application/json", "Token" => logintoken}, 
+       params: { id: schedule_window.id} 
+      expect(response).to have_http_status(200)
+     puts response.body
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: You need to validate the response body.  puts statements should be removed, as they are only used to help debug as you develop.

File: spec/requests/schedule_windows_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    end
+    let!(:organization) { FactoryBot.create(:organization) }
+    let!(:driver) { FactoryBot.create(:driver, organization_id: organization.id) }
+    let!(:location) { FactoryBot.create(:location) }
+    let!(:schedule_window) { FactoryBot.create(:schedule_window, driver_id: driver.id, location_id: location.id) }
+     it "Gets the schedule_window" do
+     get '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken},
+     params: {start_date: "02-02-2019", end_date: "05-03-2019"}
+     expect(response).to have_http_status(200)
+     puts response.body
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: You need to validate the response body and take out the puts statement.

File: spec/requests/schedule_windows_spec.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    end
+    
+    it "delete schedule_window" do 
+     token = logintoken
+     delete "/api/v1/availabilities/#{schedule_window.id}", headers: { "ACCEPT" => "application/json", "Token" => token}, 
+     params: {id: schedule_window.id }
+     expect(response).to have_http_status(200)
+      puts response.body
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: You need to validate the response body and take out the puts statement.  Validation of the response body is done by comparing the fields actually returned in that response with expected values.

File: config/routes.rb [hunk 8]
```diff
@@ -60,5 +60,9 @@
   namespace :rider do
     root :to => "welcome#rider"
   end
+  
+  namespace :admin do
+    resource :riders
+  end
```
Comment (jrmcgarvey) [inline config/routes.rb]: The changes to this file are not necessary and do not belong in this PR

File: app/views/riders/show.html.erb [hunk 9]
```diff
@@ -37,6 +37,6 @@
 <% end %>
 
 <%= link_to 'Edit', edit_rider_path(@rider) %> |
-<%= link_to 'Back', riders_path %> |
+<%= link_to 'Back', admin_riders_path %> |
```
Comment (jrmcgarvey) [inline app/views/riders/show.html.erb]: The changes to this file are unnecessary and do not belong in this PR

File: app/views/riders/index.html.erb [hunk 10]
```diff
@@ -23,9 +23,9 @@
       <td><%= rider.phone %></td>
       <td><%= rider.email %></td>
 <!--      <td><%#= rider.tokens %></td>-->
-      <td><%= link_to 'Show', rider_path(rider) %></td>
-      <td><%= link_to 'Edit', edit_rider_path(rider) %></td>
-      <td><%= link_to 'Destroy', rider_path(rider),
+      <td><%= link_to 'Show', admin_riders_path(rider) %></td>
+      <td><%= link_to 'Edit', admin_edit_rider_path(rider) %></td>
+      <td><%= link_to 'Destroy', admin_ride_path(rider),
```
Comment (jrmcgarvey) [inline app/views/riders/index.html.erb]: The changes in this file are unnecessary and do not belong in this PR

File: app/controllers/riders_controller.rb [hunk 11]
```diff
@@ -48,7 +48,7 @@ def destroy
     @rider = Rider.find(params[:id])
     @rider.destroy
 
-    redirect_to riders_path
+    redirect_to admin_riders_path
```
Comment (jrmcgarvey) [inline app/controllers/riders_controller.rb]: all this stuff about admin_riders_path needs to be taken out of this PR.  It should not be done in any case, as it is completely unnecessary

File: app/controllers/api/v1/schedule_windows.rb [hunk 12]
```diff
@@ -69,7 +69,12 @@ class ScheduleWindows < Grape::API
 
       desc "Update an schedule window for a driver"
       params do
-        requires :id, type: String, desc: "ID of avaliablity"
```
Comment (jrmcgarvey) [inline app/controllers/api/v1/schedule_windows.rb]: This line should not be deleted.

File: app/views/layouts/administration.html.erb [hunk 13]
```diff
@@ -22,7 +22,7 @@
           <%= link_to 'Drivers', drivers_path, :class => 'nav-link' %>
           </li>
           <li class="nav-item">
-              <%= link_to 'Riders', riders_path, :class => 'nav-link' %>
+              <%= link_to 'Riders', riders_path_path, :class => 'nav-link' %>
```
Comment (jrmcgarvey) [inline app/views/layouts/administration.html.erb]: This file is changed but should not be.  There are a number of other files like this, which have nothing to do with the PR and have to be restored to what they were.

File: spec/requests/schedule_windows_spec.rb [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+     expect(parsed_json ['start_time']).to eq("2019-03-11T17:00:00.000Z")
+     expect(parsed_json['end_time']).to eq("2019-03-11T21:00:00.000Z")
+     expect(parsed_json['is_recurring']).to eq(true)
+    
+     # It gets all schedule_window
+     get '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken}
+     expect(response).to have_http_status(200)
+     parsed_json = JSON.parse(response.body)
+     window_id = parsed_json['json'][0]['eventId']
+
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: Need more expect statements here, to make sure that the length and json values are correct

File: spec/requests/schedule_windows_spec.rb [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+     get '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken}
+     expect(response).to have_http_status(200)
+     parsed_json = JSON.parse(response.body)
+     window_id = parsed_json['json'][0]['eventId']
+        
+     # it gets schedule_window by ID
+      get "/api/v1/availabilities/window/#{window_id}", headers: {"ACCEPT" => "application/json", "Token" => logintoken},
+      params: { id: window_id }
+      expect(response).to have_http_status(200)
+
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: Need more expect statements here, to make sure that the length and json values are correct.

File: spec/requests/schedule_windows_spec.rb [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    end
+    let!(:organization) { FactoryBot.create(:organization) }
+    let!(:driver) { FactoryBot.create(:driver, organization_id: organization.id) }
+    let!(:location) { FactoryBot.create(:location) }
+    it "Sign up for availabilities, get all availabilities, update and delete " do 
+     post '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken}, params: { 
+         start_date: "09-01-2019",
+         end_date: "10-03-2019", 
+         start_time: "11-03-2019, 2:00 pm", 
+         end_time: "11-03-2019, 9:00 pm",
```
Comment (jrmcgarvey) [inline spec/requests/schedule_windows_spec.rb]: The validation requires that the start_date, end_date, start_time, and end_time be in the future.  The current values will start to throw errors in november.  So you need stuff like start_date = Time.now + 5.days , end_date = TIme.now + 3.months

File: spec/requests/schedule_windows_spec.rb [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    start_date = time + 3.months 
+    end_date = time + 5.months
+    start_time = time + 2.hours
+    end_time = time + 5.hours
+
+   
+    it "creates a availabilities with recurring false " do 
+     post '/api/v1/availabilities', headers: {"ACCEPT" => "application/json", "Token" => logintoken}, params: { 
+         start_date: start_date,
+         end_date: end_date,
```
[the remark this answers is not in the document]
Comment (Kwasi008) [reply spec/requests/schedule_windows_spec.rb]: Hi @micronix it does break, it seems to be a requirement on swagger, will you like for me to change that on swagger.

[the remark this answers is not in the document]
Comment (micronix) [reply spec/requests/schedule_windows_spec.rb]: I'm thinking that to make the api easier to understand if instead we only had to pass `start_time`, `end_time` and `recurring`. In the backend we should handle `start_date` and `end_date`. I think what you currently have is fine we can change that in another pull request.

File: app/controllers/api/v1/event_helpers.rb [hunk 18]
```diff
@@ -2,53 +2,6 @@ module Api
   module V1
```
Comment (micronix) [inline app/controllers/api/v1/event_helpers.rb]: Looks like we removed all the code from this file, so delete this file.

File: app/controllers/api/v1/schedule_windows.rb [hunk 19]
```diff
@@ -26,13 +26,9 @@ class ScheduleWindows < Grape::API
         optional :end, type: Time, desc: "End date for schedule"
       end
       get "availabilities", root: :schedule_windows do
-        driver = current_driver
-        schedules = ScheduleWindow.where(driver_id: driver.id)
-        all_events = []
         start_time = DateTime.now
```
Comment (micronix) [inline app/controllers/api/v1/schedule_windows.rb]: This is an API call, we are getting passed the parameters `:start` and `:end` but we are not using them. What you want to do is check if `params[:start]` is not nil then start_time is that otherwise use `DateTime.now`. You can use the `||` construct:

```ruby
start_time = params[:start] || DateTime.now
```

Similarly for the end_time it would be the same.

File: app/models/driver.rb [hunk 20]
```diff
@@ -25,9 +23,29 @@ def generate_auth_token
...
+    list = []
+    schedule_windows.each do |window|
+      list += window.events(query_start_date, query_end_date)
+    end
+    list.sort_by{|i| i[:start_time]}.reverse
   end
-
+  
+  
+  #TODO write code for non recurring pattern
```
Comment (micronix) [inline app/models/driver.rb]: Remove this commented out code for the pull request. We usually want to keep the pull request clean. I think it would be better if you use the current `recurring_weekly` method as a template instead of this commented out code.

File: app/models/recurring_pattern.rb [hunk 21]
```diff
@@ -1,4 +1,6 @@
 class RecurringPattern < ApplicationRecord
-
   belongs_to :schedule_window
+  
+  attribute :separation_count, :integer, default: 0
+  attribute :type_of_repeating, :string, default: 'weekly'
```
Comment (micronix) [inline app/models/recurring_pattern.rb]: Add a validation for the type of repeating https://guides.rubyonrails.org/active_record_validations.html#inclusion

The only possible values should be 'daily', 'weekly', 'monthly', 'yearly' I think

File: app/models/schedule_window.rb [hunk 22]
```diff
@@ -70,5 +70,73 @@ def end_date_cannot_be_before_end_time
       end
     end
   end
+  belongs_to :location
 
+  has_one :recurring_pattern
+ 
+  # after_create :create_recurring_pattern
```
Comment (micronix) [inline app/models/schedule_window.rb]: remove these comments, we will be creating the recurring pattern in the controller.

File: spec/models/schedule_window_spec.rb [hunk 23]
```diff
@@ -84,5 +70,103 @@
         subject.valid?  # run validations
         expect(subject.valid?).to be_falsey
       end
+      
+      describe 'recurring weekly' do
+        let(:recurring_pattern) { create(:recurring_weekly_pattern) }
+        
+        it "should have a weekly recurring_pattern" do
```
Comment (micronix) [inline spec/models/schedule_window_spec.rb]: change the message from "should have a weekly recurring_patter to what we are checking in this test which i think is `should return weekly events`

File: spec/models/schedule_window_spec.rb [hunk 24]
```diff
@@ -84,5 +70,103 @@
...
+          # check that start times are correct
+          start_times = events.map{|k| k[:startTime] }
+          expect(start_times).to eq(['2019-09-21 14:00', '2019-09-14 14:00'])
+          
+          # check that end times are correct
+          end_times = events.map{|k| k[:endTime] }
+          expect(end_times).to eq(['2019-09-21 16:00', '2019-09-14 16:00'])
+        end
+        
+        it "query_start_date and query_end_date can not be after start_date" do
```
Comment (micronix) [inline spec/models/schedule_window_spec.rb]: fix this message as well, it should be something along the lines of `it "should return [] when query is after the date range of the schedule window"`
